### PR TITLE
Feat: start-up/teleport recovery on networking issues

### PR DESCRIPTION
**Pull request by NickKhalow**
## What does this PR change?

The issue is described here https://github.com/decentraland/unity-explorer/issues/2821 in details

## How to test the changes?

1. Launch the explorer
2. Go to the step with the loading screen
3. Simulate an unstable internet connection (Enable/Disable it during the screen)
4. Ensure that you cannot reach a broken state in the app
5. Ensure that all features work fine as if there was no issues with the internet connection

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

### CODE REVIEW

File: Explorer/Assets/DCL/SceneLoadingScreens/LoadingScreen/ILoadingScreen.cs
Comment (dalkia) [inline Explorer/Assets/DCL/SceneLoadingScreens/LoadingScreen/ILoadingScreen.cs:11]: Why the change from `Result` to `EnumResult`? According to the [Wiki](https://github.com/decentraland/unity-explorer/wiki/Architecture-Overview#exceptions-free-async-flow), the Result contract should be used in this flows.

**Review by dalkia — CHANGES_REQUESTED** [review]

File: Explorer/Assets/DCL/PerformanceAndDiagnostics/Diagnostics/ReportsHandling/ReportHub.cs
Comment (NickKhalow) [reply Explorer/Assets/DCL/PerformanceAndDiagnostics/Diagnostics/ReportsHandling/ReportHub.cs:168]: I implemented it because we have `ReportedHub.Log` the difference is ReportedHub encapsulates ReportData and allow to non alloc closures

File: Explorer/Assets/DCL/Multiplayer/Connections/GateKeeper/Meta/SceneRoomMetaDataSource.cs
Comment (NickKhalow) [reply Explorer/Assets/DCL/Multiplayer/Connections/GateKeeper/Meta/SceneRoomMetaDataSource.cs:48]: Agreed, it's good to unify it

File: Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs
Comment (NickKhalow) [reply Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs:30]: Because I got rid off the event flow that is wonky with callbacks and I passed MinimapController directly to call its method

File: Explorer/Assets/DCL/SceneLoadingScreens/LoadingScreen/ILoadingScreen.cs
Comment (NickKhalow) [reply Explorer/Assets/DCL/SceneLoadingScreens/LoadingScreen/ILoadingScreen.cs:11]: I haven't change Wiki yet, but TaskError gives more strict info of result's state (Cancelled, Timeout and etc)

File: Explorer/Assets/Scripts/Global/Dynamic/TeleportOperations/ChangeRealmTeleportOperation.cs
Comment (NickKhalow) [reply Explorer/Assets/Scripts/Global/Dynamic/TeleportOperations/ChangeRealmTeleportOperation.cs:32]: Agreed

File: Explorer/Assets/Scripts/Global/Dynamic/MainScreenFallbackRealmNavigator.cs
Comment (NickKhalow) [reply Explorer/Assets/Scripts/Global/Dynamic/MainScreenFallbackRealmNavigator.cs:74]: Fixed

File: Explorer/Assets/DCL/SceneLoadingScreens/LoadingScreen/ILoadingScreen.cs
Comment (dalkia) [reply Explorer/Assets/DCL/SceneLoadingScreens/LoadingScreen/ILoadingScreen.cs:11]: Aren't quite the same, but with an enum to limit them? They can even be described as `Result`, as implemented in the method. 

`
public Result AsResult()
{
    if (Success)
        return Result.SuccessResult();

    var error = Error!.Value;
    return Result.ErrorResult($"{error.State}: {error.Message}");
} 
`

We even have `Result<T>`, that is also quite similar

My concern is that we are creating a divergence on how we return an operation's result. Misha has the same concern as written at the end of this [post](https://github.com/decentraland/unity-explorer/pull/2913#issue-2702173216)

The Wiki was written less than a month ago so we could all adhere to a contract; if we want to change it, we should discuss it. 

Anyways, I don't want to hold this PR because of this. When we are all back, we can discuss getting rid of the `Result` contract and using `EnumResult` moving forward. After we have a decision, we would need a cleanup PR to adhere to it.

File: Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs
Comment (dalkia) [reply Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs:38]: Can you give further details? Which event flow you got rid off?

File: Explorer/Assets/DCL/SceneLoadingScreens/LoadingScreen/ILoadingScreen.cs
Comment (NickKhalow) [reply Explorer/Assets/DCL/SceneLoadingScreens/LoadingScreen/ILoadingScreen.cs:11]: Agreed, we should unify them. I found that Result<T> is limiting in cases of task implementations when it suits good for default methods without async

File: Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs
Comment (dalkia) [reply Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs:38]: Got it, but do you need the `LazyMap`? I did a quick test on this [branch](https://github.com/decentraland/unity-explorer/commit/438c0d96f60e3326e3763a1f98c6b81053869c87).

Initializating `MapController` on the `MinimapPlugin.cs` constructor seems to work just fine.

Comment (dalkia) [reply Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs:38]: Forget it, I think I found the problem you solved. 

The issue is the cyclic dependency on `container.ChatMessagesBus`. If we do it eagerly, that reference is passed as null, and the button wont work. So, I guess that the lazy initialization is fine

File: Explorer/Assets/DCL/PerformanceAndDiagnostics/Diagnostics/ReportsHandling/ReportHub.cs
Comment (dalkia) [reply Explorer/Assets/DCL/PerformanceAndDiagnostics/Diagnostics/ReportsHandling/ReportHub.cs:168]: Cool, maybe we should discuss it and add it to the wiki as the default method. We should avoid divergence as much as we can

**Review by dalkia — DISMISSED** [review]
~Thanks for addressing the comments! Good work 🦾~

Update 16-12:
I actually detected an error, so dismissing the review

File: Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs
Comment (dalkia) [reply Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs:38]: @NickKhalow 

Actually, I was wrong. As it is, the button `Go Back to Genesis City` does not work. So, going back to the previous comment, the lazy initialization is not needed; and we have to fix the issue with the button

Comment (NickKhalow) [reply Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs:38]: Actually refactor hadn't been dedicated to `ChatMessagesBus`

It was a dependency conflict with `RealmMisc`, because it requires to directly call `MinimapController` and I wanted to avoid moving `MinimapController` instantiation to `DynamicWorldContainer`

Comment (NickKhalow) [reply Explorer/Assets/DCL/PluginSystem/Global/MinimapPlugin.cs]: I removed lazy map because it won't solve the original problem. I left it with `Inject()` hack but it should be redesigned with my proposal for DynamicWorldContainer.

just copy here for history:

_I noticed an issue in DynamicWorldContainer. It's possible to pass an uninitialized dependency to an object that requires it. For example, you might pass container.ChatMessagesBus, assuming it's initialized, but in reality, it's only initialized dozens of lines later. This can lead to a null reference error in a completely different part of the code due to the invalid dependency.
To solve this, I propose to make DynamicWorldContainer immutable and construct it only at the return steps. This approach will prevent such issues at compile time by ensuring that references cannot be passed before they are initialized._

**Review by dalkia — APPROVED** [review]

**Review by DafGreco — APPROVED** [review]
Approving this PR after checking different MB/s on Windows and Macos :

Test made on this PR:

Fake connectivity test :

**Internet connection 2MB/s test:**
teleported properly to olavra
teleported properly to metadynelabs
Teleported to Genesis plaza
Teleported to Pravus
Teleported to onboardingdcl
Emotes loaded properly
Backpack loaded properly

**Internet connection 1MB/s test:**
teleported properly to olavra
teleported properly to metadynelabs
Teleported to Genesis plaza
Teleported to Pravus
Teleported to onboardingdcl
Emotes loaded properly
Backpack loaded properly

**Limit internet to 0.5 MB/s:**
teleported properly to olavra
teleported properly to metadynelabs
Teleported to Genesis plaza
Teleported to Pravus
Teleported to onboardingdcl
Emotes loaded properly
Backpack loaded properly

Sanity check performed :
- Emotes 
- Teleport through map and chat command 
- Metadynelabs
- On boarding 
- Backpack 
- Chat interactions

**Pull request merged**
